modules/sdk/src/main/scala/org/tessellation/sdk/infrastructure/snapshot/MajorityPeerSelect.scala

```diff
@@ -42,40 +55,38 @@ object MajorityPeerSelect {
     val logger = Slf4jLogger.getLoggerFromClass[F](MajorityPeerSelect.getClass)
 
     def select: F[Peer] = selectPeer
+      .flatTap(details => logger.debug(details.asJson.noSpaces))
+      .map(_.selectedPeer)
 
-    def selectPeer: F[Peer] = storage.getResponsivePeers
-      .map(_.filter(_.state === Ready))
-      .flatMap(getPeerSublist)
-      .flatMap { peers =>
-        peers.toNel match {
-          case Some(value) => value.pure[F]
-          case None =>
-            logger.error("No Ready peers were found to be selected.") >>
-              NoPeersToSelect.raiseError[F, NonEmptyList[Peer]]
-        }
-      }
-      .flatMap(filterPeerList)
-      .map(_.toList)
-      .flatMap(Random[F].elementOf)
-
-    def filterPeerList(peers: NonEmptyList[Peer]): F[NonEmptyList[Peer]] =
-      peers
-        .parTraverseN(maxConcurrentPeerInquiries)(snapshotClient.getLatestOrdinal(_))
-        .map(_.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size })
-        .map { case (majorityOrdinal, _) => majorityOrdinal }
-        .flatMap { majorityOrdinal =>
-          peers
-            .parTraverseN(maxConcurrentPeerInquiries)(peer => getSnapshotHashByPeer(peer, majorityOrdinal).attempt)
+    def selectPeer: F[FilteredPeerDetails] = for {
+      peers <- storage.getResponsivePeers
+        .map(_.filter(_.state === Ready))
+        .flatMap(getPeerSublist)
+        .flatMap { peerSublist =>
+          MonadThrow[F].fromOption(peerSublist.toNel, NoPeersToSelect)
         }
+      latestOrdinals <- peers.parTraverseN(maxConcurrentPeerInquiries)(snapshotClient.getLatestOrdinal(_))
+      (majorityOrdinal, _) = latestOrdinals.groupBy(identity).maxBy { case (_, ordinals) => ordinals.size }
+      peerDistribution <- peers
+        .parTraverseN(maxConcurrentPeerInquiries)(peer => getSnapshotHashByPeer(peer, majorityOrdinal).attempt)
         .flatMap { peerSnapshotHashes =>
           MonadThrow[F].fromOption(
             peerSnapshotHashes.collect { case Right(peerSnapshot) => peerSnapshot }.toNel,
             NoPeersToSelect
           )
         }
         .map(_.groupMap { case (_, hash) => hash } { case (peer, _) => peer })
-        .map(_.values.toList.sortBy(-_.size))
-        .map(_.head)
+      peerCandidates = peerDistribution.values.toList.sortBy(-_.length).head
+      selectedPeer <- Random[F].elementOf(peerCandidates.toList)
+    } yield
+      FilteredPeerDetails(
+        peers,
```

Comment:
also, if I ask for `selectPeer`, why I get all the peers?
IMO, if you want to log `FilteredPeerDetails` somehow, it should be an intermediate step within the for comprehension, but the return type should be the selected peer itself. wdyt?

Reply:
For clarity, `selectPeer` has been renamed to `getFilteredPeerDetails`. The logging and the actual process of selecting the peer was separated to keep the code simpler.

Reply:
:+1: